ports/nrf/mphalport.c

```diff
@@ -57,18 +60,148 @@ void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
         NRFX_ASSERT(err);
 }
 
-void mp_hal_stdout_tx_strn_cooked(const char *str, mp_uint_t len) {
-    const char cr = '\r';
+#else
+
+#include "tusb.h"
 
-    while (len--) {
-        if (*str == '\n')
-            mp_hal_stdout_tx_strn(&cr, sizeof(cr));
+int mp_hal_stdin_rx_chr(void) {
+    for (;;) {
+        #ifdef MICROPY_VM_HOOK_LOOP
+            MICROPY_VM_HOOK_LOOP
+        #endif
+        // if (reload_requested) {
+        //     return CHAR_CTRL_D;
+        // }
 
-        mp_hal_stdout_tx_strn(str++, sizeof(char));
+        if (tud_cdc_available()) {
+            #ifdef MICROPY_HW_LED_RX
+            gpio_toggle_pin_level(MICROPY_HW_LED_RX);
+            #endif
+            return tud_cdc_read_char();
+        }
     }
+
+    return 0;
 }
-#endif
 
-void mp_hal_stdout_tx_str(const char *str) {
-    mp_hal_stdout_tx_strn(str, strlen(str));
+bool mp_hal_stdin_any(void) {
+  return tud_cdc_available() > 0;
+}
+
+void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
+
+    #ifdef MICROPY_HW_LED_TX
+    gpio_toggle_pin_level(MICROPY_HW_LED_TX);
+    #endif
+
+    #ifdef CIRCUITPY_BOOT_OUTPUT_FILE
+    if (boot_output_file != NULL) {
+        UINT bytes_written = 0;
+        f_write(boot_output_file, str, len, &bytes_written);
+    }
+    #endif
+
+    tud_cdc_write(str, len);
 }
+
+#endif // USB
+
+#endif // NUS
+
+
+/*------------------------------------------------------------------*/
+/* delay
+ *------------------------------------------------------------------*/
+void mp_hal_delay_ms(mp_uint_t delay) {
```

Comment:
No need for these delay functions, use NRFX_DELAY

Reply:
Ah thanks, I will update the delay us to use nrfx. For ms func, we need them to run the background job for usb handling, (copied from samd port)

Reply:
no worries, when it got merged, I will merge the update master to usbboot branch as well